cardano-node/src/Cardano/Node/Types.hs

```diff
@@ -92,47 +119,138 @@ newtype MaxConcurrencyDeadline = MaxConcurrencyDeadline
+type NodeDnsAddress  = NodeAddress' NodeHostDnsAddress
 
-instance FromJSON NodeAddress where
+
+instance FromJSON addr => FromJSON (NodeAddress' addr) where
   parseJSON = withObject "NodeAddress" $ \v -> do
     NodeAddress
       <$> v .: "addr"
       <*> ((fromIntegral :: Int -> PortNumber) <$> v .: "port")
 
-instance ToJSON NodeAddress where
+instance ToJSON addr => ToJSON (NodeAddress' addr) where
   toJSON na =
     object
       [ "addr" .= toJSON (naHostAddress na)
       , "port" .= (fromIntegral (naPort na) :: Int)
       ]
 
--- Embedding a Maybe inside a newtype is somewhat icky but this seems to work
--- and removing the Maybe breaks the functionality in a subtle way that is difficult
--- to diagnose.
-newtype NodeHostAddress
-  = NodeHostAddress { unNodeHostAddress :: Maybe IP }
+
+nodeIPv4ToIPAddress :: NodeIPv4Address -> NodeIPAddress
+nodeIPv4ToIPAddress = fmap nodeHostIPv4AddressToIPAddress
+
+nodeIPv6ToIPAddress :: NodeIPv6Address -> NodeIPAddress
+nodeIPv6ToIPAddress = fmap nodeHostIPv6AddressToIPAddress
+
+nodeDnsAddressToDomainAddress :: NodeDnsAddress -> DomainAddress
+nodeDnsAddressToDomainAddress NodeAddress { naHostAddress = NodeHostDnsAddress dns, naPort }
+  = DomainAddress (Text.encodeUtf8 dns) naPort
+
+nodeAddressToSockAddr :: NodeIPAddress -> SockAddr
+nodeAddressToSockAddr (NodeAddress addr port) =
+  case unNodeHostIPAddress addr of
+    IP.IPv4 ipv4 -> SockAddrInet  port   (IP.toHostAddress ipv4)
+    IP.IPv6 ipv6 -> SockAddrInet6 port 0 (IP.toHostAddress6 ipv6) 0
+
+nodeHostIPAddressToSockAddr :: NodeIPAddress -> SockAddr
+nodeHostIPAddressToSockAddr NodeAddress { naHostAddress = NodeHostIPAddress ip, naPort } =
+    case ip of
+      IPv4 ipv4 -> SockAddrInet  (fromIntegral naPort)   (IP.toHostAddress ipv4)
+      IPv6 ipv6 -> SockAddrInet6 (fromIntegral naPort) 0 (IP.toHostAddress6 ipv6) 0
+
+
+newtype NodeHostIPv4Address
+  = NodeHostIPv4Address { unNodeHostIPv4Address :: IPv4 }
+  deriving newtype Show
+  deriving (Eq, Ord)
+
+instance FromJSON NodeHostIPv4Address where
+  parseJSON (String ipStr) =
+    case readMaybe $ Text.unpack ipStr of
+      Just ip -> pure $ NodeHostIPv4Address ip
+      Nothing -> panic $ "Parsing of IPv4 failed: " <> ipStr
+  parseJSON invalid = panic $ "Parsing of IPv4 failed due to type mismatch. "
+                            <> "Encountered: " <> Text.pack (show invalid) <> "\n"
+
+instance ToJSON NodeHostIPv4Address where
+  toJSON (NodeHostIPv4Address ip) = String (Text.pack $ show ip)
+
+
+newtype NodeHostIPv6Address
+  = NodeHostIPv6Address { unNodeHostIPv6Address :: IPv6 }
   deriving newtype Show
   deriving (Eq, Ord)
 
-instance FromJSON NodeHostAddress where
+instance FromJSON NodeHostIPv6Address where
   parseJSON (String ipStr) =
     case readMaybe $ Text.unpack ipStr of
-      Just ip -> pure $ NodeHostAddress (Just ip)
+      Just ip -> pure $ NodeHostIPv6Address ip
+      Nothing -> panic $ "Parsing of IPv6 failed: " <> ipStr
```

Comment:
I know you're likely following the example of the code that already exists in this module, but I think we should `fail` instead of `panic` in the JSON parsers.

I'm actually not even sure why we're `panic`king as opposed to `fail`ing in the JSON parsers defined in this module.

Reply:
I don't know either, the question is how parse failures are presented to the user?

Reply:
Perhaps we can leave it like this for now and I (or someone else) can revise these JSON parsers in the near future.

Reply:
:+1: